File: solutions/javascript/nucleotide-count/1/nucleotide-count.js
//
// This is only a SKELETON file for the 'Nucleotide-Count' exercise. It's been provided as a
// convenience to get you started writing code faster.
//


export class NucleotideCounts {
  static parse(strand) {
    const {A, T, G, C} = NucleotideCounts.count(strand);

    return `${A} ${C} ${G} ${T}`;
  }

  static count(strand) {
    return strand.split('').reduce((nucleotideCount, nucleotide) => {
      if (nucleotide in nucleotideCount) {
        return {
          ...nucleotideCount,
          [nucleotide]: nucleotideCount[nucleotide] + 1
        }
      } else {
        throw new Error('Invalid nucleotide in strand');
      }
    }, { A: 0, T: 0, G: 0, C: 0})
  }
  
}

File: solutions/csharp/nucleotide-count/1/NucleotideCount.cs
public static class NucleotideCount
{
    public static IDictionary<char, int> Count(string sequence)
    {
        var count = new Dictionary<char, int>
        {
            { 'A', 0 },
            { 'C', 0 },
            { 'G', 0 },
            { 'T', 0 },
        };

        foreach (var nucleotide in sequence)
        {
            if (!count.ContainsKey(nucleotide))
            {
                throw new ArgumentException("Malformed sequence");
            }
            count[nucleotide]++;
        }
        
        return count;
    }
}